File: mobile/lib/models/habit.dart
class Habit {
  final int id;
  final int userId;
  final String name;
  final String? description;
  final String? unit;
  final double? targetValue;
  final String frequency;
  final String? icon;
  final bool isActive;

  Habit({
    required this.id,
    required this.userId,
    required this.name,
    this.description,
    this.unit,
    this.targetValue,
    required this.frequency,
    this.icon,
    required this.isActive,
  });

  factory Habit.fromJson(Map<String, dynamic> json) => Habit(
    id: json['id'] as int,
    userId: json['user_id'] as int,
    name: json['name'] as String,
    description: json['description'] as String?,
    unit: json['unit'] as String?,
    targetValue: json['target_value'] != null ? double.parse(json['target_value'].toString()) : null,
    frequency: json['frequency'] as String,
    icon: json['icon'] as String?,
    isActive: json['is_active'] as bool? ?? true,
  );

  Map<String, dynamic> toJson() => {
    'name': name,
    'description': description,
    'unit': unit,
    'target_value': targetValue,
    'frequency': frequency,
    'icon': icon,
  };
}

class HabitLog {
  final int id;
  final int habitId;
  final DateTime logDate;
  final double? value;
  final String? notes;

  HabitLog({
    required this.id,
    required this.habitId,
    required this.logDate,
    this.value,
    this.notes,
  });

  factory HabitLog.fromJson(Map<String, dynamic> json) => HabitLog(
    id: json['id'] as int,
    habitId: json['habit_id'] as int,
    logDate: DateTime.parse(json['log_date'] as String),
    value: json['value'] != null ? double.parse(json['value'].toString()) : null,
    notes: json['notes'] as String?,
  );
}

File: backend/app/models/habit.py
from datetime import datetime, timezone
from sqlalchemy import Column, Integer, String, Text, Date, Boolean, DateTime, Enum, DECIMAL, ForeignKey, UniqueConstraint
from sqlalchemy.orm import relationship

from app.database import Base


class Habit(Base):
    __tablename__ = "habits"

    id = Column(Integer, primary_key=True, autoincrement=True)
    user_id = Column(Integer, ForeignKey("users.id", ondelete="CASCADE"), nullable=False, index=True)

    name = Column(String(100), nullable=False)
    description = Column(Text)
    unit = Column(String(50))
    target_value = Column(DECIMAL(10, 2))

    frequency = Column(Enum("daily", "weekly", "per_meal", name="habit_frequency"), nullable=False, default="daily")
    icon = Column(String(50))
    is_active = Column(Boolean, default=True)

    created_at = Column(DateTime, default=lambda: datetime.now(timezone.utc), nullable=False)

    user = relationship("User", back_populates="habits")
    logs = relationship("HabitLog", back_populates="habit", cascade="all, delete-orphan")


class HabitLog(Base):
    __tablename__ = "habit_logs"

    id = Column(Integer, primary_key=True, autoincrement=True)
    habit_id = Column(Integer, ForeignKey("habits.id", ondelete="CASCADE"), nullable=False)
    user_id = Column(Integer, ForeignKey("users.id", ondelete="CASCADE"), nullable=False)

    log_date = Column(Date, nullable=False)
    value = Column(DECIMAL(10, 2))
    notes = Column(Text)

    created_at = Column(DateTime, default=lambda: datetime.now(timezone.utc), nullable=False)

    habit = relationship("Habit", back_populates="logs")

    __table_args__ = (UniqueConstraint("habit_id", "log_date", name="uq_habit_log"),)
